Add documentation for version 2 (#6)
laminas/laminas-mvc-middleware: docs/book/v2/routing-middleware.md

                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => new PipeSpec(
                            'container_id_for_middleware',
                            new class implements MiddlewareInterface {
                                public function process(
                                    ServerRequestInterface $request,
                                    RequestHandlerInterface $handler
                                ): ResponseInterface {
                                    return $handler->handle($request);
                                }
                            },
                            static function (
                                ServerRequestInterface $request,
                                RequestHandlerInterface $handler
                            ): ResponseInterface {
                                return $handler->handle($request);
                            },
                            'container_id_for_handler',
                        ),
                    ],
                ],
            ],
        ],
    ],

> ];

Reviewer: The whole code example should be split and each example should have its own heading.

Author: good idea, thanks!

```

Please note that if you pass instances of middleware or handlers directly, it is not compatible with config caching.

> You should sue that for development only.

Reviewer: "use" instead of "sue"

> The recommended variant ist passing container managed middleware, i.e. container service strings

Reviewer: "is" instead of "ist". _("ist" ist deutsch Sprache. 😉 )_

Author: > 
> 
> "is" instead of "ist". _("ist" ist deutsch Sprache. 😉 )_

idk, my fingers apparently type that automatically 🤷‍♂️  🙂

> like `my-awesome-middleware` or `SomeRequestHandler::class`.

Reviewer: I think the class name should be recommended because a simple string leads only to spelling errors and can not be checked via IDE.

# Routing Middleware

> Apart from PSR-15 `\Psr\Http\Server\MiddlewareInterface` or `\Psr\Http\Server\RequestHandlerInterface` middleware or

Reviewer: The same here.

service name strings which resolve to such instances, the `middleware` key in the route definition takes also an
instance of the `PipeSpec` class. 

Here are some examples how middleware can be dispatched:

## `PipeSpec` with PSR-15 Service Names

Middleware can be passed as a `PipeSpec` object with container service class-string literals.
This is the preferred method.

Assuming you have two middlewares, `\Application\Middleware\SelectLanguageMiddleware` which determines the language of
the visitor and `\Application\Middleware\AlbumFromRouteMiddleware` which fetches an `Album` object by the ID in the route
param. Both of them add this information to the `ServerRequestInterface` attributes which can be used by the

> `\Application\Handler\AlbumDetailHandler` at the end of the pipe: 

Reviewer: The same here.

```php
use Application\Handler\AlbumDetailHandler;
use Application\Middleware\AlbumFromRouteMiddlware;
use Application\Middleware\SelectLanguageMiddleware;
use Laminas\Mvc\Middleware\PipeSpec;
use Laminas\Router\Http\Segment;

return [
    'router' => [
        'routes' => [
            'album-details' => [
                'type' => Segment::class,
                'options' => [
                    'route' => '/album/:album_id',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => new PipeSpec(
                            SelectLanguageMiddleware::class,
                            AlbumFromRouteMiddelware::class,
                            AlbumDetailHandler::class
                        ),
                    ],
                ],
            ],
        ],
    ],
];
```

## `MiddlewareInterface` or `RequestHandlerInterface` Service Strings

Alternatively, if you only dispatch one class, you can pass a class-string or alias which resolves to a

> `MiddlewareInterface`, `RequestHandlerInterface` or `\Laminas\Stratigility\MiddlewarePipe`.

Reviewer: The same here.

                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => 'container_id_for_album_detail_handler',
                    ],
                ],
            ],
            'container-pipe' => [
                'type' => Literal::class,
                'options' => [
                    'route' => '/container-pipe',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => AlbumDetailMiddlewarePipe::class,
                    ],
                ],
            ],
             'container-middleware' => [
                 'type' => Literal::class,
                 'options' => [
                     'route' => '/container-middleware',
                     'defaults' => [
                         'controller' => PipeSpec::class,
                         'middleware' => SomeMiddleware::class,
                     ],
                 ],
             ],
       ],
    ],
];
```

##  Anonymous Interface Implementations

You can also pass an anonymous class which implements `MiddlewareInterface` or `RequestHandlerInterface` directly.

> This is primarily for development.

Reviewer 2: Add a note to the effect that anonymous classes can cause issues with configuration caching, which is **why** they should generally only be used in development.

                                    ServerRequestInterface $request,
                                    RequestHandlerInterface $handler
                                ): ResponseInterface {
                                    return $handler->handle($request);
                                }
                            },

                            // Closure
                            static function (
                                ServerRequestInterface $request,
                                RequestHandlerInterface $handler
                            ): ResponseInterface {
                                return $handler->handle($request);
                            },

                            // class-string of an Handler which will be resolved by the container
                            AlbumDetailHandler::class,
                        ),
                    ],
                ],
            ],
        ],
    ],
];
```

Please note that if you pass instances of middleware or handlers directly, it is not compatible with config caching. You
should use that for development only. The recommended variant is passing container managed middleware

> like `SomeRequestHandler::class`.

Reviewer 2: Make this into a note, please, so it gets called out more prominently. This is generally done by doing it in a blockquote, with a heading; I don't know if @froschdesign has other markup you can specify to call it out.

## Defining `:middleware` Placeholder in Route Definition

Although you can define a `:middleware` placeholder in your middleware route definition, we are strongly advising
against doing so. This can be utilized maliciously to invoke middleware that was never intended to be called directly.
Since middleware lookup uses the main application container, it can also be abused to cause instantiation of expensive

> services leading to a DoS vulnerability.

Reviewer 2: Question: should we maybe try and _detect_ this scenario, and raise an exception when found?

Reviewer 3: That would be out of scope tbh. Scanning route config could be added as part of some kind of tooling package. However, since Laminas is intentionally not opinionated it would be pretty hard to build the tool that would be smart enough to find bad practices.

Author: maybe we could somehow mark the feature as `@internal` and being subject to change?

> # Routing Middleware

Reviewer 3: Note: Mvc have no concept of routing(adj) middleware. It is something that exists only in mezzio.

Author: Yes, I mean the verb, as in "How to route Middleware", but that doesn't sound good...

Author: How about `Dispatching Middleware`? Or `Dispatch Middleware`?

I changed if to the former. If you're okay with it, I will also change the name of the file and nav entry. I didn't do that right away because it would mess up reviewing.

Reviewer 3: "routing middleware" is fine too, it's just I'm familiar with the other usage and my brain defaults to that meaning.

Reviewer 2: Yeah, "dispatch" makes sense here, because routing is being handled by the MVC. Let's go with "Dispatching Middleware" as the title of the document.

Author: thanks both of you! I renamed the file and nav entry and squashed the commits.

Apart from PSR-15 `Psr\Http\Server\MiddlewareInterface` or `Psr\Http\Server\RequestHandlerInterface` middleware or
service name strings which resolve to such instances, the `middleware` key in the route definition takes also an
instance of the `PipeSpec` class.

> Here are some examples how middleware can be dispatched:

Reviewer 3: I think it is important to put emphasis on the request handler behavior of the pipe:

```suggestion
Apart from PSR-15 `Psr\Http\Server\MiddlewareInterface` or `Psr\Http\Server\RequestHandlerInterface` middleware or
service name strings which resolve to such instances, the `middleware` key in the route definition takes also an
instance of the `PipeSpec` class for the middleware pipe comprising of the former.

Technically, middleware or middleware pipe defined in each individual mvc route is a complete request handler: it must produce a response and cannot delegate to a controller or middleware defined in another route.

Here are some examples how middleware can be dispatched:
```

## `PipeSpec` with PSR-15 Service Names

Middleware can be passed as a `PipeSpec` object with container service class-string literals. This is the preferred

> method.

Reviewer 3: > preferred method

I would argue that `'middleware' => SomeHandler::class,` is preferred. This is the preferred method to define the middleware pipe.

Author: You're right. I swapped the two sections and changed the wording a bit

> Assuming you have two middlewares, `Application\Middleware\SelectLanguageMiddleware` which determines the language of

Reviewer 3: Typo:
```suggestion
Assuming you have two middleware, `Application\Middleware\SelectLanguageMiddleware` which determines the language of
```

param. Both of them add this information to the `ServerRequestInterface` attributes which can be used by the
`Application\Handler\AlbumDetailHandler` at the end of the pipe:

```php
use Application\Handler\AlbumDetailHandler;
use Application\Middleware\AlbumFromRouteMiddlware;
use Application\Middleware\SelectLanguageMiddleware;
use Laminas\Mvc\Middleware\PipeSpec;
use Laminas\Router\Http\Segment;

return [
    'router' => [
        'routes' => [
            'album-details' => [
                'type' => Segment::class,
                'options' => [
                    'route' => '/album/:album_id',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => new PipeSpec(
                            SelectLanguageMiddleware::class,
                            AlbumFromRouteMiddelware::class,
                            AlbumDetailHandler::class
                        ),
                    ],
                ],
            ],
        ],
    ],
];
```

## `MiddlewareInterface` or `RequestHandlerInterface` Service Strings

> Alternatively, if you only dispatch one class, you can pass a class-string or alias which resolves to a
> `MiddlewareInterface`, `RequestHandlerInterface` or `Laminas\Stratigility\MiddlewarePipe`.

Reviewer 3: I find this confusing. Alternatively to what? I would prefer this to be a bit more explicit:

```suggestion
If you only dispatch one class, you can pass a container service class-string or alias literal which resolves to a PSR-15 `Psr\Http\Server\MiddlewareInterface` or `Psr\Http\Server\RequestHandlerInterface` middleware.

`Laminas\Stratigility\MiddlewarePipe` implements both interfaces and could be provided by the application container as a configured instance. It is what is used internally for the `PipeSpec` definitions.
```

                'type' => Literal::class,
                'options' => [
                    'route' => '/container-handler',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => 'container_id_for_album_detail_handler',
                    ],
                ],
            ],
            'container-pipe' => [
                'type' => Literal::class,
                'options' => [
                    'route' => '/container-pipe',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => AlbumDetailMiddlewarePipe::class,
                    ],
                ],
            ],
             'container-middleware' => [
                 'type' => Literal::class,
                 'options' => [
                     'route' => '/container-middleware',
                     'defaults' => [
                         'controller' => PipeSpec::class,
                         'middleware' => SomeMiddleware::class,
                     ],
                 ],
             ],
       ],
    ],
];
```

##  Anonymous Interface Implementations

> You can also pass an anonymous class which implements `MiddlewareInterface` or `RequestHandlerInterface` directly. This
> is primarily for development because it can cause issues with configuration caching.

Reviewer 3: Missing word?

```suggestion
You can also pass an anonymous class which implements `MiddlewareInterface` or `RequestHandlerInterface` directly. This
is intended primarily for development because it can cause issues with configuration caching.
```

```php
use Laminas\Diactoros\Response\TextResponse;
use Laminas\Mvc\Middleware\PipeSpec;
use Laminas\Router\Http\Literal;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\MiddlewareInterface;
use Psr\Http\Server\RequestHandlerInterface;

return [
    'router' => [
        'routes' => [
            'instance-middleware' => [
                'type' => Literal::class,
                'options' => [
                    'route' => '/instance-middleware',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => new class implements MiddlewareInterface {
                            public function process(
                                ServerRequestInterface $request,
                                RequestHandlerInterface $handler
                            ): ResponseInterface {

>                                 // You have to return a response here because attempting to use $handler here will
>                                 // always result in an empty pipeline exception.

Reviewer 3: ```suggestion
                                // You have to return a response here because attempting to use $handler here will
                                // always result in an empty pipeline exception.
                                // This middleware pipe does not delegate to controllers or middleware from other routes. 
```
